### PR TITLE
Implement Exploratory Daemon

### PR DESCRIPTION
**Pull request by jkosh44**
This commits adds the functionality to take online copies of a
postgres database. The goal of the copy is to be able to run fast
but not necessarily have the most up to date data. The process of
taking a copy is the following:
1. Run checkpoint on original database (This is not strictly
required).
2. Take a ZFS snapshot and clone of pgdata directory.
3. Run pg_resetwal on ZFS copy. This will prune the log to some
consistent point which makes starting the new instance much faster.
However it also results in some lost data.
4. Start exploratory docker container which uses the copied pgdata
directory to start a postgres instance.

The process of tearing down an exploratory instance is the
following:
1. Stop postgres instance.
2. Shutdown docker container.
3. Delete copied data.

The daemon needs to be run on a system with an existing postgres
instance and ZFS installed.

The current implementation just takes a single copy, runs some
random queries, and tears down the copy. In the future this
functionality will be daemonized to allow on demand copies to
be made.

### CODE REVIEW

File: cmudb/exploration/sql.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import os
+import subprocess
+import time
+from typing import List
+
+from pgnp_docker import execute_in_container
+from util import execute_sys_command, OutputStrategy, CONTAINER_BIN_DIR
+
+
+# SQL functionality
```
Comment (jkosh44) [inline cmudb/exploration/sql.py:10]: Based on @mkpjnx demo, may want to switch to psycopg package

File: cmudb/exploration/util.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+UTF_8 = "utf-8"
+
+
+class OutputStrategy(Enum):
+    Capture = (subprocess.PIPE, subprocess.PIPE)
+    Print = (sys.stdout, sys.stderr)
+    Hide = (subprocess.DEVNULL, subprocess.DEVNULL)
+
+
+def execute_sys_command(cmd: Union[str, List[str]],
```
Comment (jkosh44) [inline cmudb/exploration/util.py:40]: I have very mixed feelings about wrapping `subprocess.Popen`. On the one hand I find `subprocess.Popen` a bit cumbersome to use by itself. On the other hand it seems like all of our projects wrap the method in slightly different ways, and they often get a bit messy.

**Comment by lmwnshn** [comment]
For wider publicity: yes, this PR has been sitting out here,

- I am happy with the code as it is, in theory it could merge
- I am currently working on integrating it with something else to see if there are things I would change

File: cmudb/exploration/exploratory_daemon.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+from sql import checkpoint, execute_sql, \
+    wait_for_pg_ready
+from util import ZFS_DOCKER_VOLUME_POOL, REPLICA_VOLUME_POOL, REPLICA_PORT, EXPLORATION_PORT, \
+    EXPLORATION_CONTAINER_NAME, \
+    DOCKER_VOLUME_DIR, execute_sys_command
+
+
+def main():
+    """
+    The exploratory daemon is responsible for creating a copy of replica instances, to be used for model training.
```
Comment (jkosh44) [inline cmudb/exploration/exploratory_daemon.py:16]: @lmwnshn Is this clear enough or too vague? I didn't want to dump a bunch of commands to run but I also wanted to provide some detail.

File: cmudb/exploration/exploratory_daemon.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        directory path that docker uses for volumes
+    Returns
+    -------
+    exploratory_instance
+        docker process that is running exploratory instance
+    valid
+        True if the container started successfully, False otherwise
+    """
+    print("Taking checkpoint in replica")
+    checkpoint(replica_port)
```
Comment (jkosh44) [inline cmudb/exploration/exploratory_daemon.py:99]: @lmwnshn I think honestly we should remove this. I'm not really sure how much we gain by calling a checkpoint and it ends up being really expensive a lot of the time.

File: cmudb/build/pgindent.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Comment (lmwnshn) [inline cmudb/build/pgindent.sh]: Please remove this file -- I think we're giving up on `pgindent`.

**Review by lmwnshn — CHANGES_REQUESTED** [review]
LGTM pending one nit. Thanks for the documentation! will be super helpful

**Review by lmwnshn — APPROVED** [review]
LGTM.

**Pull request merged**